### ISSUE TITLE
Design bug in Kernel::Construct_base_vector_3

### ISSUE THREAD

**Issue by lrineau** [CGAL/cgal#862]
https://sympa.inria.fr/sympa/arc/cgal-develop/2015-05/msg00033.html

```
From: Marc Glisse <marc.glisse@inria.fr>
To: cgal-develop@inria.fr
Subject: Re: [cgal-develop] Design bug in Kernel::Construct_base_vector_3
Date: Sun, 17 May 2015 00:07:18 +0200 (CEST)
```
On Fri, 8 May 2015, Laurent Rineau (CGAL/GeometryFactory) wrote:
>
> While debugging one of my code (see [pr]), I found out a "bug" in the
> implementation of Construct_base_vector_3 for CGAL Cartesian kernels
> (including Epick and Epeck).
>
> The documentation of Construct_base_vector_3 is here:
>
> http://doc.cgal.org/latest/Kernel_23/classKernel_1_1ConstructBaseVector__3.html
>
> Given a `Plane_3 plane`, it returns two orthogonal vectors `b1` and `b2` so
> that (b1, b2, `plane.orthogonal_vector()`) is a direct base.
>
>
> The code is here:
> https://github.com/CGAL/cgal/blob/3fffcdb/Cartesian_kernel/include/CGAL/Cartesian/function_objects.h#L1710
>
> I quote it for you:
> ```c++
> result_type
> operator()( const Plane_3& h, int index ) const
> {
> if (index == 1) {
> if ( CGAL_NTS is_zero(h.a()) ) // parallel to x-axis
> return Vector_3(FT(1), FT(0), FT(0));
>
> if ( CGAL_NTS is_zero(h.b()) ) // parallel to y-axis
> return Vector_3(FT(0), FT(1), FT(0));
>
> if ( CGAL_NTS is_zero(h.c()) ) // parallel to z-axis
> return Vector_3(FT(0), FT(0), FT(1));
>
> return Vector_3(-h.b(), h.a(), FT(0));
> } else {
> [...]
> ```
> The code is similar of Homogeneous:
> https://github.com/CGAL/cgal/blob/3fffcdbe606f380683451194a6b59dbcdd58ca7a/Homogeneous_kernel/include/CGAL/Homogeneous/function_objects.h#L1838
>
> The code is perfectly right, if we assume an exact number type. And of course
> with an inexact number type, the construction is not exact. So the contract is
> respected.
>
>
> But I think it should be less inexact, with inexact number types.
>
> Imagine that the normal of the plane is: (epsilon_1, epsilon_2, 1.) where both
> epsilon_1 and 2 and very small numbers, compared to 1, but not equal to zero.
> In that case, the plane is almost parallel to the xy plane. But (-epsilon_2,
> epsilon_1, 0.) is returned as the first base vector, and the uncertainty of
> the direction of the vector is huge!
>
> I propose another implementation, similar to what I did in
> <CGAL/Triangulation_2_projection_traits_3.h>:
>
> https://github.com/lrineau/cgal/blob/5ea63b7b66f450a6dcf0e96b8711fdadc286d56a/Triangulation_2/include/CGAL/Triangulation_2_projection_traits_3.h#L311
> ```c++
> // n is the normal to the plane.
> const FT& nx = n.x();
> const FT& ny = n.y();
> const FT& nz = n.z();
> const FT anz = CGAL::abs(nz);
> if(anz >= CGAL::abs(nx) &&
> anz >= CGAL::abs(ny)) { // z is the max absolute coordinate of n
> b1 = Vector_3(nz, 0, -nx);
> }
> else {
> b1 = Vector_3(ny, -nx, 0);
> }
> ```
> Note that if 'z' is not the maximal absolute coordinate of 'n', then 'x' or
> 'y' is. That implementation ensures that the maximal absolute coordinate of
> the normal of the plane is used to compute the base of the plane. In the case
> I have pointed out before, the first base vector would be:
> (1., 0., -epsilon1)
> which is optimal.
>
> The proposed new code is more costly (more comparisons of coordinates), but
> will reduce uncertainty. That will be faster as soon as there are dynamic
> filters using `Interval_nt`. And it will be less inexact, of course.
>
> Comments?

Want to handle line_project_pointC2 as well while you are there? ;-)

The idea makes sense to me, although:
if (anz >= any) use nx and nz
else use nx and ny
seems simpler to me to ensure that the max is used.

Note that with Interval_nt (assuming this functions gets called from the middle of some predicate), the new comparisons may cause extra filter failures. For instance, if x, y and z are all close to 1 but the intervals don't say for sure which is biggest, you could pick any of the coordinates as "far enough from 0", but the filter will actually fail. It probably isn't that important in practice, but I am mentioning it because in NewKernel_d, I had the surprise that most filter failures came from selecting the pivot for Gauss reduction of matrices in Eigen...

--
Marc Glisse

**Comment by MaelRL** [comment]
Can the issue please be copied here so there's no need to log on there?

**Comment by lrineau** [comment]
I have modified the message at the top.

**Comment by MaelRL** [comment]
Fixed by @sgiraudot in https://github.com/CGAL/cgal/pull/2122 ?

**Comment by lrineau** [comment]
> Fixed by @sgiraudot in #2122 ?

That is right. Thanks @sgiraudot.
